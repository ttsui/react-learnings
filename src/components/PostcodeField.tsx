import React, { useEffect, useState } from "react";
import { PhoneField } from "./Fields";

export const PostcodeField: React.FC<{ value?: number; onChange: (value: number) => void }> = props => {
  const [value, setValue] = useState<number | string>("");
  const [isInvalid, setIsInvalid] = useState(false);

  useEffect(
    () => {
      setValue(props.value || "");
    },
    [props.value]
  );

  return (
    <div>
      <PhoneField
        label="Postcode"
        value={value}
        onChange={value => {
          setValue(value);

          const postcode = Number(value);
          if (!Number.isNaN(postcode)) {
            props.onChange(postcode);
            setIsInvalid(false);
          } else {
            setIsInvalid(true);
          }
        }}
      />
      {isInvalid ? <div className="t-field-validation-error">Invalid postcode</div> : null}
    </div>
  );
};
