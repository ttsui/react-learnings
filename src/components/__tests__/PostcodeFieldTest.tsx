import React from "react";
import { fireEvent, render } from "react-testing-library";
import { PostcodeField } from "../PostcodeField";

it("renders", () => {
  expect(render(<PostcodeField value={1234} onChange={() => {}} />).container).toMatchSnapshot();
});

it("calls onChange handler", () => {
  const changeHandler = jest.fn();
  const { getByLabelText } = render(<PostcodeField onChange={changeHandler} />);

  fireEvent.change(getByLabelText("Postcode"), { target: { value: "1234" } });

  expect(changeHandler).toHaveBeenCalledWith(1234);
});

it("renders validation error message", () => {
  const { container, getByLabelText } = render(<PostcodeField onChange={() => {}} />);

  fireEvent.change(getByLabelText("Postcode"), { target: { value: "abcd" } });

  expect(container.querySelector(".t-field-validation-error")).toHaveTextContent("Invalid postcode");
});
