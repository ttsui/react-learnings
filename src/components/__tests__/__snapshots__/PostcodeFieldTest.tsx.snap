// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <div>
    <label
      style="color: rgb(44, 44, 44); display: block; font-size: 1.35rem; font-family: 'Arial Narrow', sans-serif, sans-serif-condensed; margin-bottom: 1em;"
    >
      <div
        style="float: left; margin-bottom: 0.25em;"
      >
        Postcode
      </div>
      <div>
        <input
          name="postcode"
          style="border: 1px solid #d5d5d5; border-radius: 3px; font-family: Arial, sans-serif; font-size: 18px; padding: 0.5em; width: 13em;"
          type="tel"
          value="1234"
        />
      </div>
    </label>
  </div>
</div>
`;
